lib/rails/auth/x509/subject_alt_name_extension.rb

```diff
@@ -0,0 +1,49 @@
+# frozen_string_literal: true
+
+module Rails
+  module Auth
+    module X509
+      # Provides convenience methods for subjectAltName extension of X.509 certificates
+      class SubjectAltNameExtension
+        DNS_REGEX = /^DNS:/i.freeze
+        IP_REGEX  = /^IP( Address)?:/i.freeze
+        URI_REGEX = /^URI:/i.freeze
+
+        def initialize(certificate)
+          unless certificate.is_a?(OpenSSL::X509::Certificate)
+            raise TypeError, "expecting OpenSSL::X509::Certificate, got #{certificate.class}"
+          end
+
+          @extension = certificate.extensions.detect { |ext| ext.oid == "subjectAltName" }
+          @subject_alt_names = Hash.new { |h, k| h[k] = [] }
```

Comment:
Should we instead have instance variables for each type instead of using strings to key into a hash?

Reply:
Instance vars probably make more sense here. At first, I thought I'd have more SAN types. I have changed the code and I think it's simpler now.